java/daikon/dcomp/DCInstrument.java

```diff
@@ -1837,43 +1866,226 @@ void add_exit(MethodGen mg, MethodInfo mi, int method_info_index) {
+      }
+
+      // Add a tag for the return type if it is primitive
+      if ((ret_type instanceof BasicType) && (ret_type != Type.VOID)) {
+        // System.out.printf("push tag for return  type of %s%n",
+        //                   invoke.getReturnType(pool));
+        il.append(dcr_call("push_const", Type.VOID, Type.NO_ARGS));
+      }
+      il.append(invoke);
+    }
+    return il;
+  }
+
+  /**
+   * Return instructions that will discard any primitive tags corresponding to the specified
+   * arguments. An empty instruction list will be returned if there are no primitive arguments.
+   *
+   * @param arg_types argument types of target method
+   * @return instruction list that discards primitive tags from DCRuntime's per-thread comparability
+   *     data stack
+   */
+  InstructionList discard_primitive_tags(Type[] arg_types) {
+
+    InstructionList il = new InstructionList();
+    int primitive_cnt = 0;
+    for (Type arg_type : arg_types) {
+      if (arg_type instanceof BasicType) primitive_cnt++;
+    }
+    if (primitive_cnt > 0) il.append(discard_tag_code(new NOP(), primitive_cnt));
+    return il;
+  }
+
+  /**
+   * Returns whether or not the invoke target is instrumented.
+   *
+   * @param invoke instruction whose target is to be checked
+   * @param classname target class of the invoke
+   * @param method_name target method of the invoke
+   * @param arg_types argument types of target method
+   * @return true if target instrumented
+   */
+  boolean isTargetInstrumented(
+      InvokeInstruction invoke,
+      @ClassGetName String classname,
+      String method_name,
+      Type[] arg_types) {
+    boolean callee_instrumented;
+
     if (invoke instanceof INVOKEDYNAMIC) {
       // We don't instrument lambda methods.
       // BUG: BCEL doesn't know how to get classname from an INVOKEDYNAMIC instruction.
-      // debug code
-      // System.out.printf("invokedynamic NOT the classname: %s%n", invoke.getClassName(pool));
+      if (debugHandleInvoke) {
+        System.out.printf("invokedynamic NOT the classname: %s%n", classname);
+      }
+      callee_instrumented = false;
+    } else if (is_object_method(method_name, invoke.getArgumentTypes(pool))) {
       callee_instrumented = false;
     } else {
       classname = invoke.getClassName(pool);
-      callee_instrumented = callee_instrumented(classname, method_name);
+      callee_instrumented = isClassnameInstrumented(classname, method_name);
 
-      // debug code
-      // System.out.printf("callee_instrumented: %s%n", callee_instrumented);
-      // System.out.printf("invoke host: %s%n", gen.getClassName() + "." + mgen.getName());
-      // System.out.printf("invoke targ: %s%n", classname + "." + method_name);
+      if (debugHandleInvoke) {
+        System.out.printf("invoke host: %s%n", gen.getClassName() + "." + mgen.getName());
+        System.out.printf("invoke targ: %s%n", classname + "." + method_name);
+      }
 
       if (BcelUtil.javaVersion > 8) {
```

Comment:
Why only for Java versions greater than 8?

Reply:
removed test